Modularize sys-metrics graph code

DanielMwendwa opened the pull request:
Please review PR #237 first.
Fix #130 
[more info...](https://github.com/AfricasVoices/Operations-Dashboard/pull/234#issuecomment-634614806)

DanielMwendwa committed:
add avf graphs folder and graph layout script

DanielMwendwa committed:
add constructor func that accepts DOM el as an argument

DanielMwendwa committed:
function to generate layout of graphs

DanielMwendwa committed:
add setters to graph layout class

DanielMwendwa committed:
add script to generate area chart

DanielMwendwa committed:
add constructor function

DanielMwendwa committed:
function to draw graph

DanielMwendwa committed:
function to generate scales for graph

DanielMwendwa committed:
function to generate axes for graph

DanielMwendwa committed:
func to generate gridlines for graph

DanielMwendwa committed:
func to prepare tooltip

DanielMwendwa committed:
func to generate area chart

DanielMwendwa committed:
func to update area chart

DanielMwendwa committed:
func to zoom into area chart

DanielMwendwa committed:
func to add labels to area chart

DanielMwendwa committed:
Merge branch 'master' into modularize_graph_code

DanielMwendwa committed:
Merge branch 'master' into modularize_graph_code

DanielMwendwa committed:
update functions used to prepare tooltip

DanielMwendwa committed:
update function reference

DanielMwendwa committed:
move position of code blocks

DanielMwendwa committed:
added resetIdleTimeout function and updated its ref

DanielMwendwa committed:
update x scale domain appropriately

DanielMwendwa committed:
rm code that reinitialize the chart, if user double clicks

DanielMwendwa committed:
function that adds sys-metrics graphs legend

DanielMwendwa committed:
function that watches sys-metrics outage

DanielMwendwa committed:
functiion that adjusts grid lines position on chart update

DanielMwendwa committed:
update func that zooms in the chart

DanielMwendwa committed:
functions to be invoked on chart update

DanielMwendwa committed:
invoke function used to watch outage

DanielMwendwa committed:
update creation and appending of axis elements

DanielMwendwa committed:
update generation of y extent in creating scales

DanielMwendwa committed:
invoke addMetricsLegend func

DanielMwendwa committed:
add setters for setting yLimit & feature in analysis

DanielMwendwa committed:
update this reference

DanielMwendwa committed:
Merge branch 'modular-js' into modularize_graph_code

DanielMwendwa committed:
remove code used to generate sys-metrics graphs

DanielMwendwa committed:
update this reference

DanielMwendwa committed:
update date property to datetime

DanielMwendwa committed:
updated variable to 'this' variable

DanielMwendwa committed:
invoke func used to generate gridlines

DanielMwendwa committed:
correct logic of adjusting x axis on chart update

DanielMwendwa committed:
import area chart

DanielMwendwa committed:
Clear previous graphs before redrawing

DanielMwendwa committed:
plot disk usage graph

DanielMwendwa committed:
plot memory usage graph

DanielMwendwa committed:
plot cpu usage graph

DanielMwendwa committed:
add comment before code block

IsaackMwenda commented:
Please merge master into this

DanielMwendwa committed:
Merge branch 'master' into modularize_graph_code

DanielMwendwa commented:
> Please merge master into this

Updated thanks.

IsaackMwenda commented on public/avf_graphs/area.chart.js:
```diff
@@ -0,0 +1,427 @@
+import { GraphLayout } from "./graph.layout.js";
```
The number of js files is increasing and having non-descriptive folder names for them will make it hard to maintain. Please; 
- [x] store/move all js files in one `js` folder.
- [x] Rename all js files with respect to the class name e.g UI.js -> UIController.js so that we don't have to remember 2 names.
- [x] Move css file to CSS folder.

DanielMwendwa replied:
Updated thanks.

DanielMwendwa committed:
Move css file to CSS folder

DanielMwendwa committed:
update reference of css/js files

DanielMwendwa committed:
Move js files to js folder and Rename them with respect to the class name

DanielMwendwa committed:
separate js folder into two, controllers & libs folder

DanielMwendwa committed:
update reference of css/js files

DanielMwendwa commented:
Updated the folders structure

IsaackMwenda reviewed (changes requested):
We have duplicate files in different folders under the public folder. Please delete the duplicates. Also, libs and controllers are subdirectories of js folder, not standalone directories.

DanielMwendwa committed:
Delete UI.js

duplicate

DanielMwendwa committed:
Delete app.js

duplicate file

DanielMwendwa committed:
Delete coding_progess.js 

Duplicate file

DanielMwendwa committed:
Delete systems_monitoring_graphs.js

Duplicate file

DanielMwendwa committed:
Delete traffic_monitoring_graphs.js

DanielMwendwa committed:
Delete auth.js

Duplicate file

DanielMwendwa committed:
Delete firebase

Duplicate file

DanielMwendwa committed:
Delete data.js

Duplicate file

DanielMwendwa committed:
Delete scroll.js

Duplicate file

DanielMwendwa committed:
Delete styles.css

Duplicate css

DanielMwendwa committed:
Delete area.chart.js

Duplicate file

DanielMwendwa committed:
Delete graph.layout.js

Duplicate file

DanielMwendwa committed:
moved libs folder into js folder

DanielMwendwa committed:
moved controllers folder into js folder

DanielMwendwa committed:
update files references

DanielMwendwa commented:
> We have duplicate files in different folders under the public folder. Please delete the duplicates. Also, libs and controllers are subdirectories of js folder, not standalone directories.

The duplicate files don't appear on my local repo, I had moved the files but I've been able to delete them here and I have updated the repo, thanks.

DanielMwendwa committed:
remoove commented code

IsaackMwenda commented:
> > We have duplicate files in different folders under the public folder. Please delete the duplicates. Also, libs and controllers are subdirectories of js folder, not standalone directories.
> 
> The duplicate files don't appear on my local repo, I had moved the files but I've been able to delete them here and I have updated the repo, thanks.

It happens. Once you delete locally, you have to push the delete commits so that the change is updated here.

IsaackMwenda commented on public/js/controllers/traffic_graphs.js:
```diff
@@ -1,18 +1,18 @@
 // GRAPH CONTROLLER
-export class GraphController {
+export class TrafficGraphsController {
```
rename the file according to the class name in this case traffic_graphs_controller.js so that we don't have to remember two names.

DanielMwendwa replied:
Updated thanks

IsaackMwenda reviewed (approved):
This lgtm @DanielMwendwa. with minor comments to rename js files with respect to the class names.
Also, investigate the following error in the debug console before merging
![image](https://user-images.githubusercontent.com/23135322/84989882-04d6b700-b14d-11ea-81f3-b30df5daa2a5.png)

DanielMwendwa committed:
rename controller files according to the class name

DanielMwendwa committed:
rename controller files according to the class name

DanielMwendwa commented:
> This lgtm @DanielMwendwa. with minor comments to rename js files with respect to the class names.
> Also, investigate the following error in the debug console before merging
> ![image](https://user-images.githubusercontent.com/23135322/84989882-04d6b700-b14d-11ea-81f3-b30df5daa2a5.png)

The error doesn't occur that frequently, I haven't seen it recently, it's caused when one of the values used to compute the height of the bar charts is undefined for example see the statements below; 
`.attr("height", d => y_total_sent_sms_range(d[0]) - y_total_sent_sms_range(d[1]))`
`.attr("height", d => undefined - y_total_sent_sms_range(d[1]))` : This will raise the error above
The problem is why the values become undefined, maybe its internet connection, @IsaackMwenda can you check if the error still appears. For me I can't see the error.
I've also noticed, this error doesn't affect the plotting of the bar charts most of the time.

as2388 commented on public/js/controllers/UI.controller.js:
```diff
@@ -34,7 +34,7 @@ export class UIController {
         let DOMstrings = UIController.getDOMstrings(),
             script = document.createElement('script');
-        script.setAttribute('src','scroll.js');
+        script.setAttribute('src','js/libs/scroll.js'); 
```
```suggestion
        script.setAttribute('src', 'js/libs/scroll.js');
```

DanielMwendwa replied:
Updated thanks.

DanielMwendwa replied:
Updated thanks

as2388 commented on public/js/controllers/UI.controller.js:
```diff
@@ -109,7 +109,7 @@ export class UIController {
         let DOMstrings = UIController.getDOMstrings(),
             script = document.createElement('script');
-        script.setAttribute('src','scroll.js');
+        script.setAttribute('src','js/libs/scroll.js');
```
```suggestion
        script.setAttribute('src', 'js/libs/scroll.js');
```
etc.

as2388 commented on public/js/controllers/app.controller.js:
```diff
@@ -1,6 +1,6 @@
-import { UIController } from "./UI.js";
-import { DataController } from "./data.js";
-import { AuthController } from "./auth.js";
+import { UIController } from "./UI.controller.js";
```
```suggestion
import { UIController } from "./ui_controller";
```
(The '.' is a bit confusing here because . usually separates the name from the type in file names, so better to use '_')

DanielMwendwa replied:
Updated thanks

as2388 commented on public/js/controllers/app.controller.js:
```diff
@@ -23,11 +23,11 @@ class Controller {
         document.querySelector(Controller.DOMstrings.graphContainer).innerHTML = "";
-        import("./traffic_monitoring_graphs.js").then(module => {
-            module.GraphController.clearTimers();
+        import("./traffic.graphs.controller.js").then(module => {
```
Same here and everywhere else like this, prefer traffic_graphs_controller to traffic.graphs.controller

IsaackMwenda commented:
> > This lgtm @DanielMwendwa. with minor comments to rename js files with respect to the class names.
> > Also, investigate the following error in the debug console before merging
> > ![image](https://user-images.githubusercontent.com/23135322/84989882-04d6b700-b14d-11ea-81f3-b30df5daa2a5.png)
> 
> The error doesn't occur that frequently, I haven't seen it recently, it's caused when one of the values used to compute the height of the bar charts is undefined for example see the statements below;
> `.attr("height", d => y_total_sent_sms_range(d[0]) - y_total_sent_sms_range(d[1]))`
> `.attr("height", d => undefined - y_total_sent_sms_range(d[1]))` : This will raise the error above
> The problem is why the values become undefined, maybe its internet connection, @IsaackMwenda can you check if the error still appears. For me I can't see the error.
> I've also noticed, this error doesn't affect the plotting of the bar charts most of the time.

yup, it is probably an intermittent internet connection. I can't reproduce it now.

as2388 commented on public/js/controllers/systems.graphs.controller.js:
```diff
@@ -0,0 +1,107 @@
+        d3.selectAll("svg").remove();
+
+/*                         DISK UTILIZATION GRAPH
+/*      ==================================================================
```
I don't think this huge block comment is necessary now that's there not much code below, and that the variables have sensible names. I think you can get away with // Draw disk utilization graph or no comment at all. If you think you need a large block comment like this, it's a good sign that the code below should really be in its own function

DanielMwendwa replied:
Updated comments i.e., // Draw disk utilization graph. thanks

as2388 commented on public/js/controllers/systems.graphs.controller.js:
```diff
@@ -0,0 +1,107 @@
+            d.value = +d.disk_usage.used;
+        })
+
+        const area3 = new AreaChart(
```
```suggestion
        const diskUsageChart = new AreaChart(
```

as2388 commented on public/js/libs/area.chart.js:
```diff
@@ -0,0 +1,427 @@
+        let tooltipText = `${d3.timeFormat("%Y-%m-%d (%H:%M)")(d.datetime)} 
+        Value: ${d3.formatPrefix(".2", d.value)(d.value)}`;
+
+        if (this.feature == "system-metrics") {
```
It would be cleaner if you could pass these pieces of configuration in directly rather than switching on the feature/id. That's so that if we need another area chart in future, we don't need to edit various pieces of hard-coded options in the area chart implementation, but instead just pass in the new configuration.

DanielMwendwa replied:
I find it hard to pass the configuration indirectly. I had this in mind that If we needed another area chart one would not pass the feature setting of which you get an area chart without specific features that system-metrics charts have e.g outage lines.

DanielMwendwa replied:
Updated thanks.

DanielMwendwa committed:
add space after comma

Co-authored-by: Alexander Simpson <as2388@cam.ac.uk>

DanielMwendwa committed:
Add space after comma

Co-authored-by: Alexander Simpson <as2388@cam.ac.uk>

DanielMwendwa committed:
rename area3 to diskUsageChart

Co-authored-by: Alexander Simpson <as2388@cam.ac.uk>

DanielMwendwa committed:
renamed files names appropriately with undescores

DanielMwendwa committed:
update comments on graph generation

DanielMwendwa committed:
update comments on graph generation

DanielMwendwa committed:
updated variable name to be clear

DanielMwendwa committed:
increase width of svg

DanielMwendwa committed:
rm setFeature function

DanielMwendwa committed:
add func to set configurations

DanielMwendwa committed:
pass sys metrics configurations to the area chart

DanielMwendwa committed:
Sys Metrics configurations

DanielMwendwa committed:
add formatYAxisValuesAsGB function

DanielMwendwa committed:
add functions to append metric to tooltip text

DanielMwendwa committed:
add adjustSysMetricsGridlines function

DanielMwendwa committed:
pass sys metrics configurations to area chart conditionally
